rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java

```diff
@@ -72,21 +99,58 @@ public String getName() {
                 .collect(Collectors.toSet());
     }
 
+    public synchronized void removeBatch(Map<ByteArrayWrapper, byte[]> rows) {
+        for (Map.Entry<ByteArrayWrapper, byte[]> entry : rows.entrySet()) {
+            ByteArrayWrapper key = entry.getKey();
+            byte[] keyBytes = key.getData();
+            // Make this check explicit so that users know very clearly what to expect
+            Objects.requireNonNull(keyBytes);
+            delete(keyBytes);
+        }
+    }
+
+    // Assumes elements are not present. Do not delete.
+    public synchronized void addBatch(Map<ByteArrayWrapper, byte[]> rows) {
+        storage.putAll(rows);
+    }
+
     @Override
-    public synchronized void updateBatch(Map<byte[], byte[]> rows) {
-        rows.entrySet().stream().
-                forEach(entry -> storage.put(wrap(entry.getKey()), entry.getValue()));
+    public synchronized void updateBatch(Map<ByteArrayWrapper, byte[]> rows) {
+        // Hashmap does not handle null values. If there are empty values, remove them
+        for (Map.Entry<ByteArrayWrapper, byte[]> entry : rows.entrySet()) {
+            ByteArrayWrapper key = entry.getKey();
+            byte[] value = entry.getValue();
+            byte[] keyBytes = key.getData();
+
+            // Make this check explicit so that users know very clearly what to expect
+            Objects.requireNonNull(keyBytes);
+            Objects.requireNonNull(value);
+            if (keyBytes.length != 0) {
```

Comment:
This is the wrong condition. The condition should be if (value.length != 0). This was copied from updateBatchInternal() which is wrong also (but never actually executed).
This bug was already reported (in a slack conversation). 
Red flag: There are no tests for this, being this critical to consensus.
Also there is no uniform and documented way to use updateBatch to delete items: if put(x,null) IS accepted, then updateBatch with null values should also. I'm in favor of accepting null and empty as synonyms (I know some core developers don't). In any case, coherence is key.